containers.go

```diff
@@ -972,3 +973,99 @@ func FindDockerImagesByTagPrefix(imageName string) ([]docker.APIImages, error) {
 
 	return matchingImages, nil
 }
+
+// SquashImage takes a docker image with multiple layers and squashes it into
+// a single layer.  SquashImage takes advantage of the fact that docker
+// squashes layers into a single later when exported from a container
+func SquashImage(ctx context.Context, repo, tag string) error {
+	client := DockerClient()
+	squashCtx, cancel := context.WithTimeout(context.Background(), time.Minute*30)
+	defer cancel()
+
+	squashImageId, err := imageId(repo, tag)
+	if err != nil {
+		return fmt.Errorf("SquashImage: %v", err)
+	}
+
+	tar, err := imageToTar(squashCtx, squashImageId)
+	if err != nil {
+		return fmt.Errorf("SquashImage: %v", err)
+	}
+	defer os.Remove(tar)
+
+	if err = client.ImportImage(docker.ImportImageOptions{
+		Repository: repo,
+		Tag:        tag,
+		Source:     tar,
+		Context:    squashCtx,
+	}); err != nil {
+		return fmt.Errorf("SquashImage: importing image error: %v", err)
+	}
+
+	log.Infof("Squashed image: %s:%s", repo, tag)
+
+	if err := client.RemoveImageExtended(
+		squashImageId,
+		docker.RemoveImageOptions{Force: true}); err != nil {
+		log.Infof("SquashAndRemoveImage: failed to remove image (%s:%s): %v", repo, tag, err)
+	}
+
+	return nil
+}
+
+// imageToTar creates a container from an image and exports it to a tar file
+// Note: the caller is responsible for removing the tar file
+func imageToTar(ctx context.Context, imageId string) (string, error) {
+	client := DockerClient()
+	containerName := fmt.Sprintf("export_%s", strings.ReplaceAll(imageId, ":", "_"))
+
+	container, err := client.CreateContainer(docker.CreateContainerOptions{
+		Name:    containerName,
+		Config:  &docker.Config{Image: imageId},
+		Context: ctx,
+	})
+	if err != nil {
+		return "", fmt.Errorf("create container error: %v", err)
+	}
+	defer func() {
+		if err := client.RemoveContainer(docker.RemoveContainerOptions{
+			ID:      container.ID,
+			Context: ctx,
```

Comment:
This cleanup poses an interesting problem: if the following steps fail due to a cancelled context, then this remove container RPC will never get made.

The typical solution here is to ignore the parent context's deadline ([example code](https://github.com/golang/tools/blob/master/internal/xcontext/xcontext.go)) and add a fixed, short timeout.